### ISSUE TITLE
Implement Netconnectd

### ISSUE THREAD

**Issue by bryantjk** [Robo3D/Robo-iOS-Beta#26]
User should be able to manage Wi-Fi connection of the printer via the app using the Netconnectd framework.

https://github.com/Robo3D/OctoPrint-Netconnectd

**Comment by victorevector** [comment]
1. **Get current wifi settings and list of all available wifi networks** Get request to `http://<base_url>/api/plugin/netconnectd/?apikey=<apikey>` 

2. **Send commands**  Post request to `http://<base_url>/api/plugin/netconnectd/` using this schema:
    ~~~
       {
          "command": "command2",
          "some_parameter": "some_value",
          "some_optional_parameter": 2342
        }
    ~~~
    available commands: 
    ~~~
    start_ap
    stop_ap
    refresh_wifi #refreshes the wifi list 
    configure_wifi
    forget_wifi
    reset
    ~~~

## Configure wifi

post schema:

~~~
{
          "command": "configure_wifi",
          "psk": <password>,
          "ssid": <ssid>,
          "force": true 
}
~~~

**Comment by bryantjk** [comment]
Ok I implemented the first-cut of the UI for this, but am getting an unexpected result. After getting the available access points, I'm then trying to connect to one by sending the following:
URL:
    http://timeless-turing.local./api/plugin/netconnectd

Header
    X-Api-Key 6BD1441C791B471FB438A27950AD887A

Content-Type
    application/json

Parameters
    {
    "command":"configure_wifi",
    "psk":[MY PASSWORD],
    "ssid":[MY SSID]
    }

I'm not entirely sure what I'm expected to get back. I suppose the result of a success would be that the iPhone is knocked off the Printer's Hot Spot. I'm experimenting with how to handle this gracefully as effectively my networking code has sent a request and expecting a response (not to lose the connection while waiting). 

Questions I have are:

1) What happens when a given network has multiple access points with the same SSID. Does it just automatically connect to the strongest one or should I tell it the address of the strongest one?

2) If the network does not require a password, should I omit the "psk" argument, pass nil, or pass "", or is it just ignored?

**Comment by AllenMcAfee** [comment]
@bryantjk You are correct:  If the printer successfully connects to a Wifi network there would be no response other than loss of the Printer's hotspot signal.  This is a trickier one to navigate the UI since the iPhone itself will also connect to its default network.

Assuming the default network is the one the user connected the printer to, we could introduce an overlay message that pops up after the "configure_wifi" command is sent, which could read:

1. Disconnecting from printer
2. Connecting to "default SSID" 
3. Searching for printer
4. "X" printers found on this network:
       a.  display a shortened list of printers (name only) to navigate to

Ideally each message would be driven by an actual response from the phone.

In the case where the printer is not found at the end of this workflow, the message could direct the user to change wifi networks like this:

1. No printers have been found on "SSID name", do you want to change networks?
       a. Yes
       b. Close

**Comment by victorevector** [comment]
@bryantjk 

> What happens when a given network has multiple access points with the same SSID. Does it just automatically connect to the strongest one or should I tell it the address of the strongest one?

I'm not sure about this one... what do you mean by "I tell it the address of the strongest one"? Is there another marker besides SSID that you can use to dictate to which network to connect?

>If the network does not require a password, should I omit the "psk" argument, pass nil, or pass "", or is it just ignored?

Pass an empty string: ""

**Comment by bryantjk** [comment]
@victorevector for each available AP I get this info:

    {
    address = "58:6D:8F:55:51:08";
    encrypted = 1;
    quality = "-83";
    ssid = "POS_Network";
    }

I noticed at the hotel this weekend that there were several APs for the same hotel guest ssid, but the quality (signal strength) varied greatly. Obviously, we'd want connect to the best one, so I was wondering how that is ensured.

**Comment by victorevector** [comment]
@bryantjk 

Is there a way we can test this hypothesis: "netconnectd or the raspberry pi' wifi interface connects to the strongest network given an SSID...?

**Comment by victorevector** [comment]
@bryantjk I have two APs in my house that share the same SSID. I'll try to connect to both tonight and will report back on results.

**Comment by bryantjk** [comment]
I just uploaded a new build 1.0.9 (10). This has a much more robust process for handling the WiFi connection process including informing the user what's happening during each step and allowing for user to fix things if something goes wrong (incorrect wifi credentials, printer not found, etc.)

**Comment by bryantjk** [comment]
Just caught a bug. After telling the printer what network to connect to, the iOS device then connects to the default network and looks for the printer it was just talking to. If it can't find something within 10 seconds it pops up an alert giving the user the opportunity to try to configure printer wifi again. 

For build 1.0.9 (10) the timer for this 10 sec alert mistakenly wasn't cancelled when printers successfully found on the network, leading to a situation where configuration would appear successful AND the user would get an alert of the failure. This is fixed in 1.0.9 (11)
